In [1]:
import pandas as pd
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    classification_report,
    confusion_matrix
)

import joblib

In [2]:
X_train = pd.read_csv("../data/processed/X_train.csv")
X_test = pd.read_csv("../data/processed/X_test.csv")

y_train = pd.read_csv("../data/processed/y_train.csv").squeeze()
y_test = pd.read_csv("../data/processed/y_test.csv").squeeze()

In [3]:
final_model=LogisticRegression(random_state=42,max_iter=100)
final_model.fit(X_train,y_train)

LogisticRegression(random_state=42)

In [4]:
y_pred = final_model.predict(X_test)
y_prob = final_model.predict_proba(X_test)[:,1]

In [5]:
print("Accuracy :", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall   :", recall_score(y_test, y_pred))
print("F1-Score :", f1_score(y_test, y_pred))
print("ROC-AUC  :", roc_auc_score(y_test, y_prob))

Accuracy : 0.7874911158493249
Precision: 0.6205787781350482
Recall   : 0.516042780748663
F1-Score : 0.5635036496350365
ROC-AUC  : 0.8319235288940887


In [6]:
cm = confusion_matrix(y_test, y_pred)
cm

array([[915, 118],
       [181, 193]])

In [7]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.83      0.89      0.86      1033
           1       0.62      0.52      0.56       374

    accuracy                           0.79      1407
   macro avg       0.73      0.70      0.71      1407
weighted avg       0.78      0.79      0.78      1407



In [8]:
joblib.dump(
    final_model,
    "../models/logistic_regression_churn_model.pkl"
)

['../models/logistic_regression_churn_model.pkl']

In [9]:
loaded_model = joblib.load(
    "../models/logistic_regression_churn_model.pkl"
)

In [10]:
loaded_predictions = loaded_model.predict(X_test)

In [11]:
np.array_equal(y_pred, loaded_predictions)

True

In [12]:
customer = X_test.iloc[0:1]

print("Prediction:", loaded_model.predict(customer))
print("Probability:", loaded_model.predict_proba(customer))

Prediction: [0]
Probability: [[0.99424589 0.00575411]]


# Notebook Summary

In this notebook, the final model for the Telecom Customer Churn Prediction project was selected, trained, evaluated, and saved.

## Steps Completed

- Loaded the processed training and testing datasets.
- Compared Logistic Regression and Tuned Random Forest performance.
- Selected Logistic Regression as the final model based on:
  - Highest F1-score
  - Highest ROC-AUC score
  - Better recall for identifying churn customers
  - Better interpretability for business decisions
- Retrained the final model.
- Evaluated final performance using:
  - Accuracy
  - Precision
  - Recall
  - F1-score
  - ROC-AUC
  - Confusion Matrix
- Saved the trained model using Joblib.
- Loaded the saved model and verified that it produces identical predictions.
- Tested predictions on unseen customer data.

---

# Final Model Performance

| Metric | Score |
|--------|------:|
| Accuracy | 78.75% |
| Precision | 62.06% |
| Recall | 51.60% |
| F1-Score | 56.35% |
| ROC-AUC | 83.19% |

---

# Final Observations

- The model successfully identifies customers who are likely to churn while maintaining a balance between false positives and false negatives.
- Since the dataset contains moderate class imbalance, evaluation focused on metrics beyond accuracy, especially Recall, F1-score, and ROC-AUC.
- Logistic Regression was chosen over Random Forest because it provided better overall balance and generalization performance.
- Future improvements could include class weighting, SMOTE, and advanced boosting algorithms.

---

# Project Completion

This project demonstrates a complete machine learning classification workflow, from understanding customer behavior through EDA to building, optimizing, evaluating, and saving a predictive model.

The trained model can now be integrated into a customer retention system to help telecom companies identify high-risk customers and take proactive actions.